modyn/selector/internal/selector_strategies/downsampling_strategies/scheduler_downsampling_strategy.py

```diff
@@ -0,0 +1,97 @@
+from modyn.selector.internal.selector_strategies.downsampling_strategies import AbstractDownsamplingStrategy
+from modyn.selector.internal.selector_strategies.downsampling_strategies.utils import instantiate_downsampler
+
+
+class SchedulerDownsamplingStrategy(AbstractDownsamplingStrategy):
+    def __init__(self, downsampling_config: dict, maximum_keys_in_memory: int):
+        if "first_strategy" not in downsampling_config:
+            raise ValueError(
+                "If you want to use a Downsampling Scheduler, please specify the first and the second strategy."
+            )
+        first_strategy = downsampling_config["first_strategy"]
+
+        if "second_strategy" not in downsampling_config:
+            raise ValueError(
+                "If you want to use a Downsampling Scheduler, please specify the first and the second strategy."
+            )
+        second_strategy = downsampling_config["second_strategy"]
+
+        if "threshold" not in downsampling_config:
+            raise ValueError("Please specify after how many epochs you want to switch strategy.")
+        self.threshold = downsampling_config["threshold"]
+
+        if "ratio" in downsampling_config:
+            raise ValueError(
+                "SchedulerDownsamplingStrategy has no downsampling ratio. Specify the downsampling ratio"
+                "inside first_strategy and second_strategy."
+            )
+        downsampling_config["ratio"] = 100  # just to deal with super
+
+        if "sample_then_batch" in downsampling_config:
+            raise ValueError("SchedulerDownsamplingStrategy has no downsampling mode (sample_then_batch parameter).")
+        downsampling_config["sample_then_batch"] = True  # useless, just to avoid ValueErrors
+
+        # instantiate the Scheduler
+        super().__init__(downsampling_config, maximum_keys_in_memory)
+        self.first_downsampler = instantiate_downsampler(
+            {"downsampling_config": first_strategy}, maximum_keys_in_memory
+        )
+        self.second_downsampler = instantiate_downsampler(
+            {"downsampling_config": second_strategy}, maximum_keys_in_memory
+        )
+
+        self.counter = 0
+
+        # To count the elapsed epochs we can rely on how many calls were made to the functions below.
+        # Calls can be generated from two components: Supervisor (get_training_status_bar_scale)
+        # TrainerServer (everything else). SInce we have no guarantees on the order, and we want to avoid double
+        # counting, these two variables are introduced.
+        # Note that, from the trainer server, we just need to consider get_requires_remote_computation since the other
+        # calls are made sequentially
```

Comment:
This seems a bit hacky. Could we instead have a RemoteScheduledStrategy or something, that is instantiated on the trainer server, instantiates both remote downsamplers, and then switches accordingly on the trainer server? Then we would not need to keep track of this at the trainer server. 

Furthermore, does the trainer server re request which strategy to use? I thought it would only do this during setup at the beginning

Reply:
this would also eliminate the need for a getter. I think we should definitely do this a bit differently instead of hacking into these function calls to implement the scheduler. currently i think the trainer server side Remote thing would be the best solution but maybe you hav another idea